index.bs

```diff
@@ -654,6 +638,38 @@ partial interface NeuralNetworkContext {
     tensor is specified by the *newShape* argument.
 </div>
 
+### softmax ### {#api-neuralnetworkcontext-softmax}
+Compute the softmax (normalized exponential) values of the input tensor along a
+given axis.
+<script type=idl>
+partial interface NeuralNetworkContext {
+  Operand softmax(Operand x, optional long axis = -1);
+};
+</script>
+<div algorithm=softmax>
+    **Arguments:**
+        - *x*: an {{Operand}}. The input tensor.
+        - *axis*: an optional {{long}}. The axis that the softmax would be
+            performed on, with the value in the interval [0, N) where N is the
+            rank of the input tensor. The default is -1 that means the last
+            dimension.
+
+    **Returns:** an {{Operand}}. The output tensor that contains the softmax
+    results. The shape of the output tensor is the same as the shape of the
+    input tensor.
+
+    <div class="note">
+    The behavior of this operation can be generically emulated from the usage of
+    other operations as follow. However, user agents typically have a more
+    efficient implementation for it, therefore its usage is encouraged from the
+    performance standpoint.
+    <pre highlight="js">
+    return nn.div(nn.exp(x), nn.reduceSum(nn.exp(x), [axis], /* keepDimensions = */ true));
+    </pre>
```

Comment:
@wchao1115 , thanks for pointing out the incompatibility issue. The current definition actually follows the [TensorFlow `softmax`](https://www.tensorflow.org/api_docs/python/tf/nn/softmax) where the reduceSum of exponentials is only along the `axis`. However, as you mentioned, the `axis` of [ONNX `softmax`](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Softmax) "describes the axis of the input when coerced to 2D" and the reduceSum of exponentials is along axis 1 of the coerced 2D tensor. So it seems to me that the common case is only when `axis` is the last dimension of the input tensor. On the other hand, the use case of softmax for first-wave models is only for 2D input tensor. With that, I propose to only support 2D input tensor and compute reduceSum of exponentials on axis 1 for `nn.softmax` for now. What do you think?

Reply:
Another thing I need to fix is to subtract the the max value of x before compute the exponential.

Reply:
Pushed a new commit with support of the 2D input and fix of the implementation. Please take another look.

Reply:
@huningxin I can see why you want to compute the output values over the distances to the max value instead of the raw input values for numerical stability. Though this implementation trick is well known to some, it may still cause confusion to others when it doesn't seem to exactly match the canonical computation of the softmax values (even though the end result is the same mathematically) given that the description of this function also points to its canonical mathematical expression in Wikipedia.

So, my suggestion is to add some comment here to clarify *the different math* in your pseudo code, with something like "this sample deploys a well-known implementation trick to compute the exponentials of the distances to the max value, instead of the exponentials of the input values itself, in order to increase the numerical stability of the result".

I also find the explanation in the softmax paragraph of this [article](https://cs231n.github.io/linear-classify/#softmax) a worthwhile reference to this trick.

Reply:
@huningxin This is great. Please take a look at my inline comment on the updated pseudo code.

Reply:
Great point. I added a new commit to incorporate your comment. Please take another look. Thanks!